File: api/cpp/src/faas/context_impl.cpp
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2024-2024. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "api/cpp/src/faas/context_impl.h"

namespace Function {

ContextImpl::ContextImpl(std::shared_ptr<ContextInvokeParams> &invokeParams, std::shared_ptr<ContextEnv> &contextEnv)
{
    this->contextInvokeParams_ = invokeParams;
    this->contextEnv_ = contextEnv;
}

ContextImpl::ContextImpl(const ContextImpl &contextImpl)
{
    this->stateId_ = contextImpl.stateId_;
    this->state_ = contextImpl.state_;
    this->property_ = contextImpl.property_;
    this->funcStartTime_ = contextImpl.funcStartTime_;
    this->contextInvokeParams_ = contextImpl.contextInvokeParams_;
    this->contextEnv_ = contextImpl.contextEnv_;
}

ContextImpl &ContextImpl::operator=(const ContextImpl &contextImpl)
{
    this->stateId_ = contextImpl.stateId_;
    this->state_ = contextImpl.state_;
    this->property_ = contextImpl.property_;
    this->funcStartTime_ = contextImpl.funcStartTime_;
    this->contextInvokeParams_ = contextImpl.contextInvokeParams_;
    this->contextEnv_ = contextImpl.contextEnv_;
    return *this;
}

const std::string ContextImpl::GetTraceId() const
{
    return contextInvokeParams_->GetTraceId();
}

const std::string ContextImpl::GetInvokeId() const
{
    return contextInvokeParams_->GetInvokeId();
}

const FunctionLogger &ContextImpl::GetLogger()
{
    logger_.SetInvokeID(GetInvokeId());
    logger_.SetTraceID(GetTraceId());
    return logger_;
}

const std::string ContextImpl::GetInstanceId() const
{
    return stateId_;
}

const std::string ContextImpl::GetInstanceLabel() const
{
    return contextEnv_->GetInstanceLabel();
}

void ContextImpl::SetStateId(const std::string &stateId)
{
    std::lock_guard<std::mutex> lock(stateMtx_);
    this->stateId_ = stateId;
}

const std::string ContextImpl::GetState() const
{
    return state_;
}

void ContextImpl::SetState(const std::string &state)
{
    std::lock_guard<std::mutex> lock(stateMtx_);
    this->state_ = state;
}

const std::string ContextImpl::GetInvokeProperty() const
{
    return property_;
}

void ContextImpl::SetInvokeProperty(const std::string &prop)
{
    this->property_ = prop;
}

const std::string ContextImpl::GetRequestID() const
{
    return contextInvokeParams_->GetRequestId();
}

const std::string ContextImpl::GetUserData(std::string key) const
{
    return contextEnv_->GetUserData(key);
}

const std::string ContextImpl::GetFunctionName() const
{
    return contextEnv_->GetFunctionName();
}

int ContextImpl::GetRunningTimeInSeconds() const
{
    return contextEnv_->GetRunningTimeInSeconds();
}

int ContextImpl::GetRemainingTimeInMilliSeconds() const
{
    return 0;
}

const std::string ContextImpl::GetVersion() const
{
    return contextEnv_->GetVersion();
}

int ContextImpl::GetMemorySize() const
{
    return contextEnv_->GetMemorySize();
}

int ContextImpl::GetCPUNumber() const
{
    return contextEnv_->GetCPUNumber();
}

const std::string ContextImpl::GetProjectID() const
{
    return contextEnv_->GetProjectID();
}

const std::string ContextImpl::GetPackage() const
{
    return contextEnv_->GetFuncPackage();
}

const std::string ContextImpl::GetAccessKey() const
{
    return contextInvokeParams_->GetAccessKey();
}

const std::string ContextImpl::GetAlias() const
{
    return contextInvokeParams_->GetAlias();
}

const std::string ContextImpl::GetSecretKey() const
{
    return contextInvokeParams_->GetSecretKey();
}

const std::string ContextImpl::GetSecurityAccessKey() const
{
    return contextInvokeParams_->GetSecurityAccessKey();
}

const std::string ContextImpl::GetSecuritySecretKey() const
{
    return contextInvokeParams_->GetSecuritySecretKey();
}

const std::string ContextImpl::GetToken() const
{
    return contextInvokeParams_->GetToken();
}

void ContextImpl::SetFuncStartTime(const long startTime)
{
    this->funcStartTime_ = startTime;
}
}  // namespace Function

File: api/java/function-common/src/main/java/org/yuanrong/services/runtime/action/ContextImpl.java
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2024-2024. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.yuanrong.services.runtime.action;

import org.yuanrong.services.logger.UserFunctionLogger;
import org.yuanrong.services.runtime.Context;
import org.yuanrong.services.runtime.RuntimeLogger;

import com.google.gson.annotations.Expose;

import lombok.Data;

import java.util.Map;

/**
 * The type Context.
 *
 * @since 2024-07-20
 */
@Data
public class ContextImpl implements Context {
    /**
     * function logger
     */
    @Expose(serialize = false, deserialize = false)
    private UserFunctionLogger logger = new UserFunctionLogger();

    /**
     * function meta data
     */
    private FunctionMetaData funcMetaData;

    /**
     * function delegate decrypt info
     */
    private DelegateDecrypt delegateDecrypt;

    /**
     * function resource meta data
     */
    private ResourceMetaData resourceMetaData;

    /**
     * function extended meta data
     */
    private ExtendedMetaData extendedMetaData;

    /**
     * context invoke params
     */
    private ContextInvokeParams contextInvokeParams = new ContextInvokeParams();

    /**
     *  The Instance id
     */
    private String instanceID = "";

    /**
     *  The Instance Label
     */
    private String instanceLabel = "";

    /**
     * The Start time.
     */
    private long startTime = System.currentTimeMillis();

    /**
     * Stream event handler.
     */
    private Stream stream = new Stream();

    public ContextImpl() {}

    public ContextImpl(ContextImpl contextImpl) {
        if (contextImpl != null) {
            this.funcMetaData = contextImpl.getFuncMetaData();
            this.delegateDecrypt = contextImpl.getDelegateDecrypt();
            this.resourceMetaData = contextImpl.getResourceMetaData();
            this.extendedMetaData = contextImpl.getExtendedMetaData();
            this.contextInvokeParams = new ContextInvokeParams(contextImpl.getContextInvokeParams());
            this.instanceID = contextImpl.getInstanceID();
            this.instanceLabel = contextImpl.getInstanceLabel();
            this.startTime = contextImpl.getStartTime();
        }
    }

    /**
     * Set context from invoke parameters
     *
     * @param contextInvokeParams context from invoke parameters
     */
    public void setContextInvokeParams(ContextInvokeParams contextInvokeParams) {
        this.contextInvokeParams = contextInvokeParams;
    }

    /**
     * Get the extended meta data
     *
     * @return extended meta data
     */
    public ExtendedMetaData getExtendedMetaData() {
        return this.extendedMetaData;
    }

    @Override
    public String getRequestID() {
        return this.contextInvokeParams.getRequestID();
    }

    @Override
    public int getRemainingTimeInMilliSeconds() {
        return (int) (this.funcMetaData.getTimeout() * 1000L - System.currentTimeMillis() + this.startTime);
    }

    @Override
    public String getAccessKey() {
        return this.delegateDecrypt.getAccessKey();
    }

    @Override
    public String getSecretKey() {
        return this.delegateDecrypt.getSecretKey();
    }

    @Override
    public String getSecurityAccessKey() {
        return null;
    }

    @Override
    public String getSecuritySecretKey() {
        return null;
    }

    @Override
    public String getUserData(String key) {
        return this.funcMetaData.getUserData().get(key);
    }

    @Override
    public String getFunctionName() {
        return this.funcMetaData.getFuncName();
    }

    /**
     * getRunningTimeInSeconds is second in java, millSecond in other language
     *
     * @return the int
     */
    @Override
    public int getRunningTimeInSeconds() {
        return this.funcMetaData.getTimeout();
    }

    @Override
    public String getVersion() {
        return this.funcMetaData.getVersion();
    }

    @Override
    public int getMemorySize() {
        return this.resourceMetaData.getMemory();
    }

    @Override
    public int getCPUNumber() {
        return this.resourceMetaData.getCpu();
    }

    @Override
    public String getProjectID() {
        return this.funcMetaData.getTenantId();
    }

    @Override
    public String getPackage() {
        return this.funcMetaData.getService();
    }

    @Override
    public String getToken() {
        return this.delegateDecrypt.getAuthToken();
    }

    @Override
    public String getSecurityToken() {
        return this.delegateDecrypt.getSecurityToken();
    }

    @Override
    public String getAlias() {
        return this.funcMetaData.getAlias();
    }

    @Override
    public RuntimeLogger getLogger() {
        return logger;
    }

    @Override
    public Object getState() {
        return null;
    }

    @Override
    public void setState(Object state) {}

    @Override
    public String getInstanceID() {
        return this.instanceID;
    }

    @Override
    public String getInstanceLabel() {
        return this.instanceLabel;
    }

    public void setInstanceLabel(String instanceLabel) {
        this.instanceLabel = instanceLabel;
    }

    @Override
    public String getInvokeProperty() {
        return null;
    }

    @Override
    public String getTraceID() {
        return this.contextInvokeParams.getRequestID();
    }

    @Override
    public void setTraceID(String requestID) {
        this.contextInvokeParams.setRequestID(requestID);
    }

    @Override
    public String getInvokeID() {
        return this.contextInvokeParams.getInvokeID();
    }

    @Override
    public String getWorkflowID() {
        return null;
    }

    @Override
    public String getWorkflowRunID() {
        return null;
    }

    @Override
    public String getWorkflowStateID() {
        return null;
    }

    @Override
    public String getReqStreamName() {
        return null;
    }

    @Override
    public String getRespStreamName() {
        return null;
    }

    @Override
    public String getFrontendResponseStreamName() {
        return null;
    }

    @Override
    public String getIAMToken() {
        return "";
    }

    @Override
    public Map<String, String> getExtraMap() {
        return null;
    }

    @Override
    public void setExtraMap(Map<String, String> extraMap) {
    }

    @Override
    public Stream getStream() {
        stream.setRequestId(this.getInvokeID());
        stream.setInstanceId(this.getInstanceID());
        return this.stream;
    }
}
